### PR TITLE
Move the Spring Cloud Config server to vert.x instead of Apache HTTP client

### PR DESCRIPTION
**Pull request by aureamunoz**
Similar work than the done in the [PR #14600](https://github.com/quarkusio/quarkus/pull/14600).

Could you, please, take a look @geoand ? Thanks!!!!

### CODE REVIEW

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/VertxSpringCloudConfigGateway.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                try {
+                    return OBJECT_MAPPER.readValue(bodyAsString, Response.class);
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException("Got unexpected error " + e.getOriginalMessage());
+                }
+            }
+        });
+    }
+
+    //    @Override
```
Comment (gastaldi) [inline extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/VertxSpringCloudConfigGateway.java:214]: Better remove this commented out code completely

Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/VertxSpringCloudConfigGateway.java:214]: Oups! thank you :-)

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java [hunk 2]
```diff
@@ -28,23 +32,33 @@ public SpringCloudConfigServerClientConfigSourceProvider(SpringCloudConfigClient
...
         this.activeProfile = activeProfile;
 
-        springCloudConfigClientGateway = new DefaultSpringCloudConfigClientGateway(springCloudConfigClientConfig);
+        springCloudConfigClientGateway = new VertxSpringCloudConfigGateway(springCloudConfigClientConfig);
     }
 
     @Override
     public Iterable<ConfigSource> getConfigSources(ClassLoader forClassLoader) {
+        List<ConfigSource> result = new ArrayList<>();
+        List<Uni<?>> allUnis = new ArrayList<>();
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:41]: Looks like we only have one `Uni`, right?

If so there is no need to use a list nor to do a `combine` at the end of the method

File: extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java [hunk 3]
```diff
@@ -45,26 +47,32 @@ void testBasicExchange() throws Exception {
...
-                assertThat(ps.getSource()).contains(entry("my.prop", "from application-dev.yml"));
-            });
-            assertThat(r.getPropertySources().get(2)).satisfies(ps -> {
-                assertThat(ps.getSource()).contains(entry("foo", "from foo props"),
-                        entry("democonfigclient.message", "hello spring io"));
-            });
-            assertThat(r.getPropertySources().get(3)).satisfies(ps -> {
-                assertThat(ps.getSource()).contains(entry("foo", "baz"));
-            });
+        sut.exchange(applicationName, profile).invoke(new Consumer<Response>() {
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java:51]: I think it's best to use `.await()` and then just assert the result as was done in the existing code

**Comment by aureamunoz** [comment]
I have re factored the code and use the `await()` method as suggested. I think it's better now, could you please take a look @geoand ?

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java [hunk 4]
```diff
@@ -28,23 +29,25 @@ public SpringCloudConfigServerClientConfigSourceProvider(SpringCloudConfigClient
...
-            final Response response = springCloudConfigClientGateway.exchange(applicationName, activeProfile);
-            final List<Response.PropertySource> propertySources = response.getPropertySources();
-            Collections.reverse(propertySources); // reverse the property sources so we can increment the ordinal from lower priority to higher
-            final List<ConfigSource> result = new ArrayList<>(propertySources.size());
-            for (int i = 0; i < propertySources.size(); i++) {
-                final Response.PropertySource propertySource = propertySources.get(i);
-                // Property sources obtained from Spring Cloud Config are expected to have a higher priority than even system properties
-                // 400 is the ordinal of SysPropConfigSource, so we use 450 here
-                result.add(new InMemoryConfigSource(450 + i, propertySource.getName(), propertySource.getSource()));
+            Response response = springCloudConfigClientGateway.exchange(applicationName, activeProfile).await().atMost(springCloudConfigClientConfig.readTimeout);
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:39]: The wait timeout should not be only the read timeout here - see what I used in Consul

File: extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java [hunk 5]
```diff
@@ -78,7 +78,7 @@ private static SpringCloudConfigClientConfig configForTesting() {
         springCloudConfigClientConfig.url = "http://localhost:" + MOCK_SERVER_PORT;
         springCloudConfigClientConfig.label = Optional.of("master");
         springCloudConfigClientConfig.connectionTimeout = Duration.ZERO;
-        springCloudConfigClientConfig.readTimeout = Duration.ZERO;
+        springCloudConfigClientConfig.readTimeout = Duration.ofSeconds(10);
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java:81]: Why did you change this?

Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java:81]: ah, my fault, it was during the tests, I will undo it!

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java (hunk 4 above)
Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:39]: ah, yes :-S

**Comment by aureamunoz** [comment]
I've change it for addint more time to the `await` method, @geoand

File: extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java (hunk 5 above)
Comment (geoand) [reply extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java:83]: It doesn't seem like you undid this one :)

Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java]: today is not my day. It should be ok now. Thanks!! :-)

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java [hunk 6]
```diff
@@ -28,23 +28,26 @@ public SpringCloudConfigServerClientConfigSourceProvider(SpringCloudConfigClient
...
-            final List<Response.PropertySource> propertySources = response.getPropertySources();
-            Collections.reverse(propertySources); // reverse the property sources so we can increment the ordinal from lower priority to higher
-            final List<ConfigSource> result = new ArrayList<>(propertySources.size());
-            for (int i = 0; i < propertySources.size(); i++) {
-                final Response.PropertySource propertySource = propertySources.get(i);
-                // Property sources obtained from Spring Cloud Config are expected to have a higher priority than even system properties
-                // 400 is the ordinal of SysPropConfigSource, so we use 450 here
-                result.add(new InMemoryConfigSource(450 + i, propertySource.getName(), propertySource.getSource()));
+            Response response = springCloudConfigClientGateway.exchange(applicationName, activeProfile).await()
+                    .atMost(springCloudConfigClientConfig.readTimeout);
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:39]: I thought you changed it here as well, but I guess you only changed it in the test :)

We'll need the change here as well

Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:39]: weird, I thought I had changed it here also.

Comment (aureamunoz) [reply extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java]: ah, yes, I have seen. I should have changed it in the test instead in the provider class. Now (I think) it's ok

**Review by geoand — APPROVED** [review]
LGTM, thanks!

File: extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java [hunk 7]
```diff
@@ -28,23 +28,27 @@ public SpringCloudConfigServerClientConfigSourceProvider(SpringCloudConfigClient
...
-            final List<Response.PropertySource> propertySources = response.getPropertySources();
-            Collections.reverse(propertySources); // reverse the property sources so we can increment the ordinal from lower priority to higher
-            final List<ConfigSource> result = new ArrayList<>(propertySources.size());
-            for (int i = 0; i < propertySources.size(); i++) {
-                final Response.PropertySource propertySource = propertySources.get(i);
-                // Property sources obtained from Spring Cloud Config are expected to have a higher priority than even system properties
-                // 400 is the ordinal of SysPropConfigSource, so we use 450 here
-                result.add(new InMemoryConfigSource(450 + i, propertySource.getName(), propertySource.getSource()));
+            Response response = springCloudConfigClientGateway.exchange(applicationName, activeProfile).await()
+                    .atMost(springCloudConfigClientConfig.connectionTimeout
```
Comment (geoand) [inline extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigServerClientConfigSourceProvider.java:39]: As discussed in chat, we need to use `.await().indefinitely()`  when `connectionTimeout` and `readTimeout` are `0`

**Comment by gsmet** [comment]
Apparently, it still needs adjustments, I killed CI and marked it as draft.

**Comment by aureamunoz** [comment]
It should be Ok now.

**Comment by geoand** [comment]
@aureamunoz please cancel CI.

We are facing pressure on CI and need the resources for tomorrow's release. You can restart the jobs tomorrow when  the release is done.

Thanks

**Comment by aureamunoz** [comment]
ah, I'm sorry, I didn't know. I think I'm not able to cancel it

**Comment by geoand** [comment]
OK, I'll cancel it

**Comment by geoand** [comment]
Remind me to restart it tomorrow please

**Comment by aureamunoz** [comment]
Can we run the CI now @geoand ?

**Comment by geoand** [comment]
Yes.

I suggest rebasing the PR onto master and force pushing it. That will result in CI being executed

**Comment by aureamunoz** [comment]
Ok, yes!
